package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -407,314 +179,172 @@ cdef class NSResults(object):
-        """
-
-        if cellid < 0:
-            return False
-        if cellid >= self.size:
-            return False
-
-        cellxyz[ZZ] = <ns_int> (cellid / self.cell_offsets[ZZ])
-        cellid -= cellxyz[ZZ] * self.cell_offsets[ZZ]
-
-        cellxyz[YY] = <ns_int> (cellid / self.cell_offsets[YY])
-        cellxyz[XX] = cellid - cellxyz[YY] * self.cell_offsets[YY]
-
-        return True
-
-    cdef fill_grid(self, real[:, ::1] coords):
-        """Sorts atoms into cells based on their position in the brick shaped box
-
-        Every atom inside the brick shaped box is assigned a
-        cell-id based on its position. Another list ``beadids``
-        sort the atom-ids in each cell.
-
-        Note
-        ----
-        The method fails if any coordinate is outside the brick shaped box.
-
-        """
-
-        cdef ns_int i, cellindex = -1
-        cdef ns_int ncoords = coords.shape[0]
-        cdef ns_int[:] beadcounts = np.empty(self.size, dtype=int)
-
-        with nogil:
-            # Initialize buffers
-            for i in range(self.size):
-                beadcounts[i] = 0
-
-            # First loop: find cellindex for each bead
-            for i in range(ncoords):
-                cellindex = self.coord2cellid(&coords[i, 0])
-
-                self.nbeads[cellindex] += 1
-                self.cellids[i] = cellindex
-
-                if self.nbeads[cellindex] > self.nbeads_per_cell:
-                    self.nbeads_per_cell = self.nbeads[cellindex]
-
-        # Allocate memory
-        self.beadids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size * self.nbeads_per_cell)  # np.empty((self.size, nbeads_max), dtype=np.int)
-        if not self.beadids:
-            raise MemoryError("Could not allocate memory for _NSGrid.beadids ({} bits requested)".format(sizeof(ns_int) * self.size * self.nbeads_per_cell))
-
-        with nogil:
-            # Second loop: fill grid
-            for i in range(ncoords):
-
-                # Add bead to grid cell
-                cellindex = self.cellids[i]
-                self.beadids[cellindex * self.nbeads_per_cell + beadcounts[cellindex]] = i
-                beadcounts[cellindex] += 1
+        # TODO: Check optimising for orthogonal cells
+        cdef int xyz[3]
+
+        self.coord2cellxyz(coord, xyz)
+        
+        return xyz[0] + xyz[1] * self.cell_offsets[1] + xyz[2] * self.cell_offsets[2]
+
+    cdef void coord2cellxyz(self, const float* coord, int* xyz) nogil:
+        """Calculate cell coordinate for coord"""
+        cdef int cx, cy, cz
+        
+        cz = <int> (coord[2] / self.cellsize[2])
+        cy = <int> ((coord[1] - cz * self.triclinic_dimensions[7]) / self.cellsize[1])
+        cx = <int> ((coord[0] - cy * self.triclinic_dimensions[3] - cz * self.triclinic_dimensions[6]) / self.cellsize[0])
+        
+        xyz[0] = cx
+        xyz[1] = cy
+        xyz[2] = cz
+
+    cdef int cellxyz2cellid(self, int cx, int cy, int cz) nogil:
```

Comment:
With PBC the branching in this function should better be replaced by modulo operations, i.e.,
```python
cx %= self.ncells[0]
cy %= self.ncells[1]
cz %= self.ncells[2]
```
For non-periodic cells, there either need to be bound checks within this function or before it is called.
Also, we must be aware of the fact that the number of cells in any dimension may be less than 3, which currently leads to double-counting cells! However, this problem could probably be better dealt with in a different place.

Reply:
It's probably more correct as modulo math but I know that I'm calling this with `-1 <= cx <= ncells`.

WRT the minimum number of cells, this now done in `__init__` where there's either >3 or 1 cells in each dimension

The bounds check is done by returning `END (-1)` to give a NULL result

Reply:
> I know that I'm calling this with -1 <= cx <= ncells.

Would it be worthwhile to add `assert` here for anything you know to be true?